pkg/driver/trace_driver.go

```diff
@@ -276,10 +198,18 @@ func (d *Driver) functionsDriver(list *list.List, announceFunctionDone *sync.Wai
 	var previousIATSum int64
 
 	for {
-		if minuteIndex >= totalTraceDuration {
+		if minuteIndex != currentMinute {
+			// postpone summation of invocation count for the beginning of each minute
+			invokedSinceExperimentStarted += function.Specification.PerMinuteCount[currentMinute]
+			currentMinute = minuteIndex
+		}
+
+		iatIndex := invokedSinceExperimentStarted + invocationIndex
+
+		if minuteIndex >= totalTraceDuration || iatIndex >= len(IAT) {
 			// Check whether the end of trace has been reached
 			break
-		} else if function.InvocationStats.Invocations[minuteIndex] == 0 {
+		} else if function.Specification.PerMinuteCount[minuteIndex] == 0 {
 			// Sleep for a minute if there are no invocations
```

Comment:
I think that is not needed now, we cover empty minutes by having IATs>60s. I don't think we need to sleep in addition to that.

Reply:
I think we should add a test for the empty minute case. In this test, we might check the interval between invocation with the intended interval for several cases in test mode. This would act as proof that we don't drift too much from the generated IATs.

Reply:
I think the individual function driver should be rewritten from scratch. Have a look at the last commit where I did that. What do you think about clock drift here? Previously, we were doing adjustments on a minute-basis, now we do it on a per-invocation basis.